tt-train/sources/ttml/core/indestructible.hpp

```diff
@@ -0,0 +1,38 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <type_traits>
+#include <utility>
+
+namespace ttml::core {
+
+template <typename T>
+class Indestructible {
+public:
+    // Create (or retrieve) the single indestructible instance of T.
+    // This constructs T only once and returns a reference.
+    template <typename... Args>
+    static T& get_instance(Args&&... args) {
+        static bool constructed = false;
+        static typename std::aligned_storage<sizeof(T), alignof(T)>::type storage;
+        if (!constructed) {
```

Comment:
Is it possible to initialize the static via a lambda and not have this boolean?

Reply:
good point

Reply:
I rewrote it.